File: algorithms/1046.last-stone-weight/solution.rb
# @param {Integer[]} stones
# @return {Integer}
def last_stone_weight(stones)
  loop do
    return 0 if stones.length == 0
    return stones[0] if stones.length == 1

    stones.sort!
    l, m = stones.pop, stones.pop
    diff = l - m
    stones.push(diff) unless diff == 0
  end
end

File: algorithms/1046.last-stone-weight/solution.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn last_stone_weight(stones: Vec<i32>) -> i32 {
        let mut heap: BinaryHeap<i32> = BinaryHeap::from(stones);

        while heap.len() > 1 {
            let diff = heap.pop().unwrap() - heap.pop().unwrap();

            if diff > 0 {
                heap.push(diff);
            }
        }

        heap.pop().unwrap_or(1)
    }
}
